core/trino-main/src/main/java/io/trino/operator/TableScanWorkProcessorOperator.java

```diff
@@ -174,7 +180,11 @@ public void close()
                 source = new EmptyPageSource();
             }
             else {
-                source = pageSourceProvider.createPageSource(session, split, table, columns, dynamicFilter);
+                ConnectorPageSource pageSource = pageSourceProvider.createPageSource(session, split, table, columns, dynamicFilter);
+                synchronized (this) {
```

Comment:
This should check that cancellation did not happen. Otherwise there could be a race condition.